sys/psa_crypto/psa_persistent_storage/psa_crypto_cbor_encoder.c

```diff
@@ -0,0 +1,625 @@
+/*
+ * Copyright (C) 2023 HAW Hamburg
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     sys_psa_crypto sys_psa_crypto_slot_mgmt
+ * @{
+ *
+ * @file
+ * @brief       API to encode PSA Crypto keys to CBOR for persistent storage
+ *
+ * @author      Lena Boeckmann <lena.boeckmann@haw-hamburg.de>
+ *
+ * @}
+ */
+
+#include "nanocbor/nanocbor.h"
+#include "psa_crypto_slot_management.h"
+
+#define ENABLE_DEBUG    0
+#include "debug.h"
+
+/**
+ * @brief   Convert a nanocbor error to a string for debugging
+ *
+ * @param   res Nanocbor error value
+ * @return  const char*
+ */
+static const char *nanocbor_error_to_string(int res)
+{
+    switch (res) {
+        case NANOCBOR_ERR_OVERFLOW:
+            return "NANOCBOR_ERR_OVERFLOW";
+        case NANOCBOR_ERR_INVALID_TYPE:
+            return "NANOCBOR_ERR_INVALID_TYPE";
+        case NANOCBOR_ERR_END:
+            return "NANOCBOR_ERR_END";
+        case NANOCBOR_ERR_RECURSION:
+            return "NANOCBOR_ERR_RECURSION";
+        case NANOCBOR_NOT_FOUND:
+            return "NANOCBOR_NOT_FOUND";
+        default:
+            return "Error value not recognized";
+    }
+}
+
+/**
+ * @brief   Convert nanocbor error to PSA Crypto status value
+ *
+ * @param   res Nanocbor error value
+ * @return  psa_status_t
+ */
+static psa_status_t nanocbor_error_to_psa_status(int res)
+{
+    switch (res) {
+        case NANOCBOR_ERR_END:
+            return PSA_ERROR_BUFFER_TOO_SMALL;
+        case NANOCBOR_NOT_FOUND:
+            return PSA_ERROR_DOES_NOT_EXIST;
+        default:
+            return PSA_ERROR_GENERIC_ERROR;
+    }
+}
+
+/**
+ * @brief   Encode key data stored in a PSA key slot in CBOR.
+ *
+ * @param   enc             Encode object of type @c nanocbor_encoder_t
+ * @param   slot            Key slot containing the key to encode
+ * @param   is_public_key   True if key to encode is asymmetric public key,
+ *                          false if it's a single key or asymmetric private key
+ *
+ * @return  psa_status_t
+ */
+static psa_status_t psa_encode_key_data(nanocbor_encoder_t *enc, uint8_t *key_data,
+                                        size_t key_data_len)
+{
+    int res;
+
+    res = nanocbor_fmt_uint(enc, key_data_len);
+    if (res < 0) {
+        goto error;
+    }
+
+    res = nanocbor_fmt_array(enc, key_data_len);
```

Comment:
Couldn't we get red of the extra uint as the key data length is already encoded in the array length?